### PR TITLE
construct rules in envoy Rbac config for policy dark launch 

### PR DESCRIPTION
**Pull request by quanjielin**
https://github.com/istio/istio/issues/5345 

pilot side's change to distribute two rules(enforced/permissive) to envoy config.

### CODE REVIEW

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 1]
```diff
@@ -199,18 +197,53 @@ func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindin
...
+			continue
+		}
 
-		if len(permissions) != 0 {
+		enforcedPrincipals, permissivePrincipals := convertToPrincipals(roleToBinding[role.Name])
+		principals := make([]*policyproto.Principal, 0)
+		principals = append(principals, enforcedPrincipals...)
+		principals = append(principals, permissivePrincipals...)
+
+		if len(principals) == 0 {
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: Why move this from line 186 to 210? If len(principals) == 0 we could return early and no need to call `convertToPermission()`.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 2]
```diff
@@ -199,18 +197,53 @@ func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindin
...
 			// Constructs the policy for a role with both permissions and bindings.
-			rbac.Policies[role.Name] = &policyproto.Policy{
+			permissiveRbac.Policies[role.Name] = &policyproto.Policy{
 				Permissions: permissions,
 				Principals:  principals,
 			}
-			log.Debugf("role %v generated policy: %v", role.Name, *rbac.Policies[role.Name])
+			log.Debugf("role %v generated permissive policy: %v", role.Name, *permissiveRbac.Policies[role.Name])
+
+			continue
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: Please use `else` instead of `continue` to handle the ENFORCED case.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 3]
```diff
@@ -199,18 +197,53 @@ func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindin
...
+		}
+
+		// If ServiceRole is in ENFORCED mode, ServiceRoleBindings that
+		// refers to it will decide their own enforcement modes.
+		rbac.Policies[role.Name] = &policyproto.Policy{
+			Permissions: permissions,
+			Principals:  enforcedPrincipals,
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: There could be problems here, if `len(enforcedPrincipals) == 0`, it violates the [validation rule](https://github.com/envoyproxy/envoy/blob/49c940f6f23daf2d53943d4311dec2e5a4ccd1dd/api/envoy/config/rbac/v2alpha/rbac.proto#L82) of principals

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 4]
```diff
@@ -199,18 +197,53 @@ func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindin
...
 		}
+		log.Debugf("role %v generated enforced policy: %v", role.Name, *rbac.Policies[role.Name])
+
+		permissiveRbac.Policies[role.Name] = &policyproto.Policy{
+			Permissions: permissions,
+			Principals:  permissivePrincipals,
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: Same here.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 5]
```diff
@@ -250,14 +283,23 @@ func convertToPermission(rule *rbacproto.AccessRule) *policyproto.Permission {
 }
 
 // convertToPrincipals converts a list of subjects to principals.
-func convertToPrincipals(bindings []*rbacproto.ServiceRoleBinding) []*policyproto.Principal {
+func convertToPrincipals(bindings []*rbacproto.ServiceRoleBinding) ([]*policyproto.Principal, []*policyproto.Principal) {
 	principals := make([]*policyproto.Principal, 0)
+	permissivePrincipals := make([]*policyproto.Principal, 0)
 	for _, binding := range bindings {
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: The logic inside this for-loop seems a little redundant, could you refactor?
Some points
- the `continue` in line 294 just skip the next `for` statement in line 296, replace it to `else` would increase the readability
- line 291-293 and line 296-298 are doing the almost the same thing, could you merge them into one loop
- please also update the comment

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 4 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: added check

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 1 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: call convertToPrincipals or convertToPermission first looks the same to me(neither of them is heavy)..

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 5 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: updated, for 2 since we check binding.mode before the loop and the logic is simple, I prefer to keep it straightforward.

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 6]
```diff
@@ -529,3 +529,278 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+								},
+							},
+						},
+					},
+				},
+			},
+		}},
+	}
+
+	permissivePolicy1 := &policy.Policy{
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: Can you add comment and/or simple function to generate this policy, so it's easier to understand what is the expected output.
Alternative, in the actual test comparison, compare the key component with golden value.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 2 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: done.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 3 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: added check.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 7]
```diff
@@ -249,15 +284,24 @@ func convertToPermission(rule *rbacproto.AccessRule) *policyproto.Permission {
 	return &policyproto.Permission{Rule: rules}
 }
 
-// convertToPrincipals converts a list of subjects to principals.
-func convertToPrincipals(bindings []*rbacproto.ServiceRoleBinding) []*policyproto.Principal {
-	principals := make([]*policyproto.Principal, 0)
+// convertToPrincipals converts lists of subjects to principals(enforced/permissive mode).
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: What is the definition of enforced/permissive principals. If the binding rules is Enforced, but bind to role that is Permissive, what is the principals mode then?

File: pilot/pkg/networking/plugin/authz/rbac_test.go (hunk 6 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: updated to use helper functions.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 7 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: updated comments, 'mode' should be only used to servicerole/binding now.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 1 above)
Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/rbac.go]: The stringMatch in line 194 and the TODO to add more matching behavior makes it more expensive to check the permissions first.

Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: maybe, not worth arguing on this anyway :) updated.

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 8]
```diff
@@ -529,3 +529,190 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+			"service-role-1": permissivePolicy1,
+			"service-role-2": permissivePolicy2,
+		},
+	}
+	testCases := []struct {
+		name           string
+		service        string
+		rbac           *policy.RBAC
+		permissiveRbac *policy.RBAC
+	}{
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: May be add test for corner case, e.g empty roles, empty bindings etc. It doesn't really make sense to have this array but have only one test.

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 9]
```diff
@@ -529,3 +529,190 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+		Policies: map[string]*policy.Policy{
+			"service-role-1": permissivePolicy1,
+			"service-role-2": permissivePolicy2,
+		},
+	}
+	testCases := []struct {
+		name           string
+		service        string
+		rbac           *policy.RBAC
+		permissiveRbac *policy.RBAC
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: probably keep the name consistent with output, e.g shadowRbac.

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 10]
```diff
@@ -529,3 +529,190 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
 		}
 	}
 }
+
+func TestConvertRbacRulesToFilterConfigPermissive(t *testing.T) {
+	roles := []model.Config{
+		{
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: This can be a function too.

File: pilot/pkg/networking/plugin/authz/rbac_test.go (hunk 8 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: added.

File: pilot/pkg/networking/plugin/authz/rbac_test.go (hunk 9 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: renamed.

File: pilot/pkg/networking/plugin/authz/rbac_test.go (hunk 10 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: done.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 11]
```diff
@@ -180,16 +180,24 @@ func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindin
...
+	permissiveRbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
 	for _, role := range roles {
-		log.Debugf("checking role %v for service %v", role.Name, service)
-		principals := convertToPrincipals(roleToBinding[role.Name])
+		enforcedPrincipals, permissivePrincipals := convertToPrincipals(roleToBinding[role.Name])
+		principals := make([]*policyproto.Principal, 0)
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: Why don't you have the  convert function above the return the full list, based on role.Mode and binding mode?

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 12]
```diff
@@ -199,18 +207,43 @@ func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindin
...
 			}
 		}
+		if len(permissions) == 0 {
+			log.Debugf("role %v skipped for no permissions found", role.Name)
+			continue
+		}
 
-		if len(permissions) != 0 {
+		// If ServiceRole is in PERMISSIVE mode, all ServiceRoleBindings that refers to it will be treated as PERMISSIVE mode.
+		if role.Spec.(*rbacproto.ServiceRole).Mode == rbacproto.EnforcementMode_PERMISSIVE {
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: you shouldn't need this if, if convertToPrincipal already take care of it.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 11 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: push the logic down to convertToPrincipals

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 12 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: push the logic down to convertToPrincipals

**Comment by liminw** [comment]
In general, having "permissive mode" in both roles and rolebindings really confuses me. I think it makes more sense just having "permissive mode" in rolebindings, because rolebindings are the actual policies that take effect. Roles are just for grouping permissions. If you do that, the code will be more straightforward and easy to understand.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 13]
```diff
@@ -180,16 +180,15 @@ func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindin
...
-		principals := convertToPrincipals(roleToBinding[role.Name])
-		if len(principals) == 0 {
-			// Skip to next role if found no bindings for current role. This means nobody could
-			// access the current role, so we don't need to check the remaining rules anymore.
-			log.Debugf("role %v skipped for no bindings found", role.Name)
-			continue
-		}
+		enforcedPrincipals, permissivePrincipals := convertToPrincipals(role.Spec.(*rbacproto.ServiceRole).Mode, roleToBinding[role.Name])
 
+		log.Debugf("checking role %v for service %v", role.Name, service)
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: [optional] Add
```
if checklen(enforcedPrincipals) ==0 && len(permissivePrincipals) == 0 {
  continue
}
```
for early exit if the lists are empty.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 14]
```diff
@@ -249,15 +261,39 @@ func convertToPermission(rule *rbacproto.AccessRule) *policyproto.Permission {
...
+
+		return enforcedPrincipals, permissivePrincipals
+	}
+
+	// If ServiceRole is in ENFORCED mode, ServiceRoleBindings that
+	// refers to it will decide their own enforcement modes.
 	for _, binding := range bindings {
-		for _, subject := range binding.Subjects {
-			principals = append(principals, convertToPrincipal(subject))
+		if binding.Mode == rbacproto.EnforcementMode_ENFORCED {
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: [optional] You can do this and get rid of the loop above.
```
if serviceRoleMode == rbacproto.EnforcementMode_PERMISSIVE || binding.Mode == rbacproto.EnforcementMode_ENFORCED {
...
}
```

**Comment by diemtvu** [comment]
Actually, I've just realized the RoleBinding can bind to only one role (roleref), so having just permissive mode under binding as Limin suggested should be enough.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 13 above)
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac.go]: done

**Comment by quanjielin** [comment]
one benefit of having mode no serviceRole is user don't need to specify mode for all serviceBindings that refer to that role(since serviceRole.mode override), but that needs more careful documentation though.

anyway, will remove mode from serviceRole from api definition in another PR.

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 15]
```diff
@@ -529,3 +529,187 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+	}
+
+	for _, tc := range testCases {
+		rbac := convertRbacRulesToFilterConfig(tc.service, tc.roles, tc.bindings)
+		if len(tc.roles) == 0 || len(tc.bindings) == 0 {
+			if got, want := len(rbac.Rules.Policies), 0; got != want {
+				t.Errorf("len(rbac.Rules.Policies), got: %d, want: %d", got, want)
+			}
+		} else {
+			if !reflect.DeepEqual(*tc.rbac, *rbac.Rules) {
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: Is there any part of rbac you don't want to compare? In other words, why don't use one RBAC golden data that have both rules and shadow?

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 16]
```diff
@@ -529,3 +529,187 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+		Action: policy.RBAC_ALLOW,
+		Policies: map[string]*policy.Policy{
+			"service-role-1": enforcePolicy1,
+			"service-role-2": enforcePolicy2,
+		},
+	}
+	shadowRbac := &policy.RBAC{
+		Action: policy.RBAC_ALLOW,
+		Policies: map[string]*policy.Policy{
+			"service-role-1": permissivePolicy1,
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: [optional] I would define these policies and permission inline here, to avoid adding too many variables, also quicker to cross check.

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 17]
```diff
@@ -529,3 +529,187 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+			service:  "service",
+			roles:    roles,
+			bindings: []model.Config{},
+		},
+	}
+
+	for _, tc := range testCases {
+		rbac := convertRbacRulesToFilterConfig(tc.service, tc.roles, tc.bindings)
+		if len(tc.roles) == 0 || len(tc.bindings) == 0 {
+			if got, want := len(rbac.Rules.Policies), 0; got != want {
```
[the remark this answers is not in the document]
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: our internal go unit test uses that 'got/want pattern' a lot :)

[the remark this answers is not in the document]
Comment (diemtvu) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: but with want come from test case. I find this comparison is just weird, at least from c++ perspective :). Btw, can you set `config` for this test case to an empty RBAC so you don't have to do this special treatment.
Also, talking about it, you may want to name it as `want` instead :)

[the remark this answers is not in the document]
Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: no special treatment looks better :) thanks !

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 18]
```diff
@@ -529,3 +529,183 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+			Policies: map[string]*policy.Policy{},
+		},
+	}
+
+	testCases := []struct {
+		name     string
+		service  string
+		roles    []model.Config
+		bindings []model.Config
+		config   *rbacconfig.RBAC
```
Comment (diemtvu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: config -> want or expectConfig

Comment (quanjielin) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: renamed.
